cpp/src/join/semi_join.cu

```diff
@@ -92,12 +92,22 @@ std::unique_ptr<rmm::device_uvector<cudf::size_type>> left_semi_anti_join(
                                                 equality_build);
   auto hash_table     = *hash_table_ptr;
 
-  thrust::for_each_n(rmm::exec_policy(stream),
-                     thrust::make_counting_iterator<size_type>(0),
-                     right_num_rows,
-                     [hash_table] __device__(size_type idx) mutable {
-                       hash_table.insert(thrust::make_pair(idx, true));
-                     });
+  // if compare_nulls == NOT_EQUAL, we can simply ignore any rows that are
```

Comment:
Could we also adjust the hash table size based on the number of valid rows? Not sure how important this is for join operations, just a thought.

Reply:
Something like this seems to work.  I have to do a `count_unset_bits()` on the aggregate bitmask though.  This also feels mildly spooky to change as I'm trying to treat lightly in this unfamiliar module :) 

```
auto const row_bitmask = (compare_nulls == null_equality::EQUAL)
                           ? rmm::device_buffer{}
                           : cudf::detail::bitmask_and(right_flattened_keys, stream);
size_type right_row_ignore_count = compare_nulls == null_equality::EQUAL ? 0 : 
  count_unset_bits(static_cast<const bitmask_type*>(row_bitmask.data()), 0, right_num_rows);
...
size_t const hash_table_size = compute_hash_table_size(right_num_rows - right_row_ignore_count);
```